### ISSUE TITLE
[virtualbox@macOS] Purge-uninstall leaves VMs behind

### ISSUE THREAD

**Issue by ricab** [canonical/multipass#3571]
**Describe the bug**
VMs remain registered in VirtualBox after running the uninstall script and requesting that all data and VMs be removed. The reason is that the uninstall script needs to be run as root, but root is usually not authenticated with the Multipass daemon and so cannot ask to remove VMs.

The deletion command fails with QEMU too, but removing the directories is enough to get rid of the VMs in that case.

**To Reproduce**

```
$ /Library/Application\ Support/com.canonical.multipass/uninstall.sh
-bash: /Library/Application Support/com.canonical.multipass/uninstall.sh: Permission denied
$ bash /Library/Application\ Support/com.canonical.multipass/uninstall.sh
This script needs to run as root
$ sudo bash /Library/Application\ Support/com.canonical.multipass/uninstall.sh
Are you sure you want to remove Multipass from your system? [Y/N] y
Do you want to delete all your Multipass VMs and daemon data too? [Y/N] y
Removing VMs:
delete failed: The client is not authenticated with the Multipass service.
Please use 'multipass authenticate' before proceeding.
Failed to delete multipass VMs from underlying driver
[...]
```

**Additional info**
 - OS: macOS 14
- `multipass version`: 1.13.1
- `multipass get local.driver`: `virtualbox`

**Additional context**
We should probably revisit the uninstall method altogether. That will probably be necessary for future work.

**Comment by vicdeveloperr** [comment]
I'm working on this.

**Comment by SouravVerma-art** [comment]
I'd like to work on this issue. I've reviewed the problem and have a good understanding of what needs to be done. Could you please assign this to me?

**Comment by tobe2098** [comment]
Hello @SouravVerma-art, thank you for your insterest! I will assign this to you now. Post questions here if you face any problems!

**Comment by tobe2098** [comment]
Hi @SouravVerma-art, have you found a solution?

**Comment by 7echkilla** [comment]
Hi @ricab  and @tobe2098, I have spent some time analyzing the issue and wanted to share my thoughts and get some feedback. 

At the core, it appears to be a mismatch between OS-level privilege and multipass' auth model. The macOS uninstall script runs as root, but multipass' CLI operations require an authenticated user session to communicate with the daemon. As a result, VM deletion fails during uninstall with an authentication error.

While this is currently macOS-triggered, the underlying concern seems potentially cross-platform: uninstall relies on authenticated CLI operations, which may not be valid in elevated or non-user session contexts.

The difference in backend (driver) behavior also contributes to the symptom:
- **VirtualBox** stores VM registrations in global state and requires explicit unregister operations
- **QEMU**-based setups are more fs-oriented, so partial cleanup *may* still appear successful even when CLI deletion fails

From an architectural perspective, I see two potential directions:
1. **Daemon-side cleanup**: Introduce a daemon-controlled shutdown cleanup path (or privileged cleanup entrypoint) that ensures instance and driver state are cleaned up independent of auth state during uninstall.
2. **Decouple uninstall from CLI-based deletion**: Have the uninstall process avoid relying entirely on `multipass delete` and instead use daemon shutdown behavior and/or driver-level cleanup abstractions to ensure deterministic cleanup across platforms.

High-level flow I had in mind:
```bash
uninstall -> stop multipassd -> daemon shutdown cleanup (delete instance, cleanup driver state) -> remove app components
```

In the short term, a trivial solution could be to make VM cleanup best-effort in the macOS uninstall script so that uninstall is not blocked by auth failures, while still attempting cleanup when possible.

Would it make sense for me to pick it up since there hasn't been any recent activity? I can put together a small PR for the macOS script change if this direction is acceptable, and perhaps open a follow-up issue for the longer-term architectural improvement.

Happy to adjust my understanding if I've missed anything.

**Comment by ricab** [comment]
Hi @7echkilla, thank you for your analysis. You're right about the mismatch, but the daemon can't just delete data on shutdown (shutdown ≠ uninstall). Any procedure we added to remove data would have the same issue: we'd need to authenticate the initiator, lest anyone on the machine delete all the data. Root should probably be allowed by default, but that is not the current situation and fixing it is beyond the scope of this issue.

We have been discussing better privilege models, but for the time being we could improve the script to communicate existing constraints better. Perhaps even verify authentication status first and ask the user to manually delete all instances (or authenticate as root).

The VirtualBox backend is slated for deprecation soon and I am not sure how useful this would be for the upcoming Apple Virtualization backend, so it is probably not worth spending too much time on. A simple clarification of the user interaction would be nice though.

**Comment by rawmeat21** [comment]
Hi, is anyone working on this rn? If not, I would like to be assigned to this issue.

**Comment by tobe2098** [comment]
Hi @rawmeat21, could you explain a bit how do you plan to solve the issue? Thank you :D

**Comment by rawmeat21** [comment]
> Hi [@rawmeat21](https://github.com/rawmeat21), could you explain a bit how do you plan to solve the issue? Thank you :D

From my understanding, the uninstall script can't delete the files from VirtualBox if the root is not authenticated with the multipass daemon. So why not, during uninstallation create this auth token if it doesn't exist? Then we can easily delete all files cleanly. Will that work or is there a problem in this approach?
